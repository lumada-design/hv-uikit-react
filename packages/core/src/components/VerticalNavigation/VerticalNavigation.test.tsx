import { describe, expect, it, vi } from "vitest";
import { render } from "@testing-library/react";
import { useMemo, useState } from "react";
import { LogOut, User } from "@hitachivantara/uikit-react-icons";
import userEvent from "@testing-library/user-event";
import {
  HvVerticalNavigation,
  HvVerticalNavigationAction,
  HvVerticalNavigationActions,
  HvVerticalNavigationHeader,
  HvVerticalNavigationTree,
} from "@core/components";

const Sample = ({
  collapsedMode,
  slider,
}: {
  collapsedMode?: "simple" | "icon";
  slider?: boolean;
}) => {
  const navigationData = useMemo(
    () => [
      { id: "00", label: "Overview" },
      { id: "01", label: "Analytics", selectable: false },
      {
        id: "02",
        label: "Storage",
        data: [
          {
            id: "02-01",
            label: "Cloud",
            data: [
              {
                id: "02-01-01",
                label: "Servers",
                href: "https://www.hitachivantara.com/en-us/news.html",
              },
              {
                id: "02-01-02",
                label: "HCP Anywhere",
                href: "/?path=/story/structure-vertical-navigation--main",
              },
              {
                id: "02-01-03",
                label: "This Computer",
                disabled: true,
              },
            ],
          },
        ],
      },
      {
        id: "03",
        label: "Administration",
        data: [
          {
            id: "03-01",
            label: "Rest API",
            data: [
              {
                id: "03-01-01",
                label: "Log Bundle",
              },
            ],
          },
        ],
      },
    ],
    []
  );

  const [value, setValue] = useState("00");

  const [show, setShow] = useState(true);

  const handleExpand = () => {
    setShow(!show);
  };
  return (
    <div style={{ display: "flex", width: 220, height: 530 }}>
      <HvVerticalNavigation
        id="sample1"
        collapsedMode={collapsedMode}
        slider={slider}
      >
        <HvVerticalNavigationHeader
          title="Menu"
          onCollapseButtonClick={handleExpand}
          collapseButtonProps={{
            "aria-label": "collapseButton",
            "aria-expanded": show,
          }}
        />
        <HvVerticalNavigationTree
          aria-label="Example 1 navigation"
          collapsible
          selected={value}
          onChange={(event, data) => {
            console.log(data);
            if (data.id === "02-01-01") {
              event.preventDefault();
              event.stopPropagation();
            }
            setValue(data.id);
          }}
          data={navigationData}
        />
        <HvVerticalNavigationActions>
          <HvVerticalNavigationAction label="Profile" icon={<User />} />
          <HvVerticalNavigationAction label="Logout" icon={<LogOut />} />
        </HvVerticalNavigationActions>
      </HvVerticalNavigation>
    </div>
  );
};

const SliderSample = () => {
  const navigationData = useMemo(
    () => [
      {
        id: "menu1",
        label: "Menu 1",
        path: "",
        data: [
          {
            id: "menu1-1",
            label: "Menu 1-1",
            path: "",
            parent: null,
          },
          {
            id: "menu1-2",
            label: "Menu 1-2",
            path: "",
            data: [
              {
                id: "menu1-2-1",
                label: "Menu 1-2-1",
                path: "",
                parent: null,
              },
              {
                id: "menu1-2-2",
                label: "Menu 1-2-2",
                path: "",
                parent: null,
              },
              {
                id: "menu1-2-3",
                label: "Menu 1-2-3",
                path: "",
                parent: null,
              },
            ],
            parent: null,
          },
          {
            id: "menu1-3",
            label: "Menu 1-3",
            path: "",
            parent: null,
          },
        ],
        parent: null,
      },
      {
        id: "menu2",
        label: "Menu 2",
        path: "",
        parent: null,
      },
      {
        id: "menu3",
        label: "Menu 3",
        path: "",
        parent: null,
      },
    ],
    []
  );

  const [value, setValue] = useState("00");

  return (
    <div style={{ display: "flex", width: 220, height: 530 }}>
      <HvVerticalNavigation id="sample1" slider>
        <HvVerticalNavigationHeader
          title="Menu"
          backButtonProps={{ "aria-label": "backButton" }}
        />
        <HvVerticalNavigationTree
          aria-label="Example 1 navigation"
          collapsible
          selected={value}
          onChange={(event, data) => {
            console.log(data);
            if (data.id === "02-01-01") {
              event.preventDefault();
              event.stopPropagation();
            }
            setValue(data.id);
          }}
          data={navigationData}
        />
      </HvVerticalNavigation>
    </div>
  );
};

describe("VerticalNavigation", () => {
  const { ResizeObserver } = window;

  beforeEach(() => {
    // @ts-ignore
    delete window.ResizeObserver;
    window.ResizeObserver = vi.fn().mockImplementation(() => ({
      observe: vi.fn(),
      unobserve: vi.fn(),
      disconnect: vi.fn(),
    }));
  });

  afterEach(() => {
    window.ResizeObserver = ResizeObserver;
    vi.restoreAllMocks();
  });

  it("should be defined", () => {
    const { container } = render(<Sample />);
    expect(container).toBeDefined();
  });

  it("should render correctly", () => {
    const { container } = render(<Sample />);
    expect(container).toMatchSnapshot();
  });

  it("should collapsed", async () => {
    const { getByRole, getByLabelText } = render(
      <Sample collapsedMode="simple" />
    );

    const collapseButton = getByRole("button", { name: "collapseButton" });
    expect(collapseButton).toBeInTheDocument();
    expect(collapseButton).toHaveAttribute("aria-expanded", "true");

    const nav = getByLabelText("Example 1 navigation");
    expect(nav).toHaveStyle(`display : block`);

    await userEvent.click(collapseButton);

    expect(collapseButton).toHaveAttribute("aria-expanded", "false");
    expect(nav).not.toHaveStyle(`display : none`);
  });

  it("should have icons", async () => {
    const { getByRole, getByLabelText } = render(
      <Sample collapsedMode="icon" />
    );

    const collapseButton = getByRole("button", { name: "collapseButton" });
    expect(collapseButton).toBeInTheDocument();
    expect(collapseButton).toHaveAttribute("aria-expanded", "true");

    const nav = getByLabelText("Example 1 navigation");
    expect(nav).toHaveStyle(`display : block`);

    await userEvent.click(collapseButton);

    expect(collapseButton).toHaveAttribute("aria-expanded", "false");
    expect(nav).toHaveStyle(`display : block`);
  });

  describe("Slider Navigation", async () => {
    it("should change header text", async () => {
      const { getByText, getByRole, queryByText } = render(<SliderSample />);

      const title = queryByText("Menu");
      expect(title).not.toBeInTheDocument();

      const navigateButton = getByRole("button");
      expect(navigateButton).toBeInTheDocument();

      await userEvent.click(navigateButton);

      const newTitle = getByText("Menu 1");
      expect(newTitle).toBeInTheDocument();

      expect(queryByText("Menu 2")).not.toBeInTheDocument();
    });

    it("should navigate to child", async () => {
      const { getByText, getByRole, queryByText, getByLabelText } = render(
        <SliderSample />
      );

      const title = queryByText("Menu");
      expect(title).not.toBeInTheDocument();

      const navigateButton = getByRole("button");
      expect(navigateButton).toBeInTheDocument();

      await userEvent.click(navigateButton);

      const newTitle = getByText("Menu 1");
      expect(newTitle).toBeInTheDocument();

      expect(queryByText("Menu 2")).not.toBeInTheDocument();

      const goBackButton = getByLabelText("backButton");
      expect(goBackButton).toBeInTheDocument();

      await userEvent.click(goBackButton);

      expect(queryByText("Menu 2")).toBeInTheDocument();
    });
  });
});
