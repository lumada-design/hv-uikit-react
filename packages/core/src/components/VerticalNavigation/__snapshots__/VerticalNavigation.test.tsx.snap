// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`VerticalNavigation > should render correctly 1`] = `
<div>
  <div
    style="display: flex; width: 220px; height: 530px;"
  >
    <div
      class="HvVerticalNavigation-root css-1400ftt-StyledRoot eyka99t0"
      id="sample1"
    >
      <div
        class="HvVerticalNavigationHeader-root css-12dbsib-StyledHeader e1arrhkt0"
      >
        <h3
          class="HvTypography-root HvTypography-title3 css-ywsy37-getStyledComponent e1tnpalo0"
        >
          Menu
        </h3>
        <button
          aria-expanded="true"
          aria-label="collapseButton"
          class="HvButton-root css-1910alx-StyledButton e138pvrm0"
        >
          <div
            class="css-1nxr552-StyledContentDiv e138pvrm3"
          >
            <span
              class="css-1o4qkkz-StyledChildren e138pvrm1"
            >
              <div
                class="css-f95wjz"
                name="Backwards"
              >
                <svg
                  focusable="false"
                  height="16"
                  viewBox="0 0 16 16"
                  width="16"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <path
                    class="color0"
                    d="M11.30863 15.5 4 8 11.30863.5 12 1.30536 5.38272 8 12 14.79529z"
                    fill="var(--uikit-colors-acce1)"
                  />
                </svg>
              </div>
            </span>
          </div>
        </button>
      </div>
      <nav
        aria-label="Example 1 navigation"
        class="HvVerticalNavigationTree-root css-ovxg28-StyledNav ejose0b0"
      >
        <ul
          class="HvVerticalNavigationTreeView-root HvVerticalNavigationTree-list css-1xsm1gc-StyledRoot e7gq1cg0"
        >
          <li
            class="HvVerticalNavigationTreeViewItem-node HvVerticalNavigationTreeViewItem-selectable HvVerticalNavigationTreeViewItem-selected css-1y9s9wc-StyledNode er1eqbm1"
            id="hvtreeview2-00"
          >
            <div
              aria-current="page"
              class="HvVerticalNavigationTreeViewItem-content er1eqbm0 HvTypography-root HvTypography-body css-1yu9rnx-getStyledComponent-StyledContent e1tnpalo0"
              id="hvtreeview2-00-button"
              role="button"
              style="padding-left: 10px;"
              tabindex="0"
            >
              <div
                class=""
                data-mui-internal-clone-element="true"
                style="overflow: hidden; text-overflow: ellipsis; white-space: nowrap;"
              >
                Overview
              </div>
            </div>
          </li>
          <li
            class="HvVerticalNavigationTreeViewItem-node HvVerticalNavigationTreeViewItem-unselectable css-1y9s9wc-StyledNode er1eqbm1"
            id="hvtreeview2-01"
          >
            <div
              class="HvVerticalNavigationTreeViewItem-content er1eqbm0 HvTypography-root HvTypography-body css-1yu9rnx-getStyledComponent-StyledContent e1tnpalo0"
              id="hvtreeview2-01-button"
              role="button"
              style="padding-left: 10px;"
              tabindex="-1"
            >
              <div
                class=""
                data-mui-internal-clone-element="true"
                style="overflow: hidden; text-overflow: ellipsis; white-space: nowrap;"
              >
                Analytics
              </div>
            </div>
          </li>
          <li
            class="HvVerticalNavigationTreeViewItem-node HvVerticalNavigationTreeViewItem-expandable HvVerticalNavigationTreeViewItem-collapsed HvVerticalNavigationTreeViewItem-unselectable css-1y9s9wc-StyledNode er1eqbm1"
            id="hvtreeview2-02"
          >
            <div
              aria-controls="hvtreeview2-02-group"
              aria-expanded="false"
              class="HvVerticalNavigationTreeViewItem-content er1eqbm0 HvTypography-root HvTypography-body css-1yu9rnx-getStyledComponent-StyledContent e1tnpalo0"
              id="hvtreeview2-02-button"
              role="button"
              style="padding-left: 0px;"
              tabindex="0"
            >
              <div
                class="css-f95wjz"
                name="DropDownXS"
              >
                <svg
                  focusable="false"
                  height="12"
                  viewBox="0 0 12 12"
                  width="12"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <path
                    class="color0"
                    d="M5.99994 8.65 1.4 4.05l.7-.7L6 7.25l3.9-3.9.7.7z"
                    fill="var(--uikit-colors-acce1)"
                  />
                </svg>
              </div>
              <div
                class=""
                data-mui-internal-clone-element="true"
                style="overflow: hidden; text-overflow: ellipsis; white-space: nowrap;"
              >
                Storage
              </div>
            </div>
            <ul
              class="HvVerticalNavigationTreeViewItem-group css-zinev9-StyledGroup er1eqbm2"
              id="hvtreeview2-02-group"
            >
              <li
                class="HvVerticalNavigationTreeViewItem-node HvVerticalNavigationTreeViewItem-expandable HvVerticalNavigationTreeViewItem-collapsed HvVerticalNavigationTreeViewItem-unselectable css-1y9s9wc-StyledNode er1eqbm1"
                id="hvtreeview2-02-01"
              >
                <div
                  aria-controls="hvtreeview2-02-01-group"
                  aria-expanded="false"
                  class="HvVerticalNavigationTreeViewItem-content er1eqbm0 HvTypography-root HvTypography-body css-1yu9rnx-getStyledComponent-StyledContent e1tnpalo0"
                  id="hvtreeview2-02-01-button"
                  role="button"
                  style="padding-left: 32px;"
                  tabindex="0"
                >
                  <div
                    class="css-f95wjz"
                    name="DropDownXS"
                  >
                    <svg
                      focusable="false"
                      height="12"
                      viewBox="0 0 12 12"
                      width="12"
                      xmlns="http://www.w3.org/2000/svg"
                    >
                      <path
                        class="color0"
                        d="M5.99994 8.65 1.4 4.05l.7-.7L6 7.25l3.9-3.9.7.7z"
                        fill="var(--uikit-colors-acce1)"
                      />
                    </svg>
                  </div>
                  <div
                    class=""
                    data-mui-internal-clone-element="true"
                    style="overflow: hidden; text-overflow: ellipsis; white-space: nowrap;"
                  >
                    Cloud
                  </div>
                </div>
                <ul
                  class="HvVerticalNavigationTreeViewItem-group css-zinev9-StyledGroup er1eqbm2"
                  id="hvtreeview2-02-01-group"
                >
                  <li
                    class="HvVerticalNavigationTreeViewItem-node HvVerticalNavigationTreeViewItem-selectable HvVerticalNavigationTreeViewItem-unselected css-1y9s9wc-StyledNode er1eqbm1"
                    id="hvtreeview2-02-01-01"
                  >
                    <a
                      class="HvVerticalNavigationTreeViewItem-content HvVerticalNavigationTreeViewItem-link er1eqbm0 HvTypography-root HvTypography-body css-1yu9rnx-getStyledComponent-StyledContent e1tnpalo0"
                      href="https://www.hitachivantara.com/en-us/news.html"
                      id="hvtreeview2-02-01-01-button"
                      role="button"
                      style="padding-left: 74px;"
                      tabindex="0"
                    >
                      <div
                        class=""
                        data-mui-internal-clone-element="true"
                        style="overflow: hidden; text-overflow: ellipsis; white-space: nowrap;"
                      >
                        Servers
                      </div>
                    </a>
                  </li>
                  <li
                    class="HvVerticalNavigationTreeViewItem-node HvVerticalNavigationTreeViewItem-selectable HvVerticalNavigationTreeViewItem-unselected css-1y9s9wc-StyledNode er1eqbm1"
                    id="hvtreeview2-02-01-02"
                  >
                    <a
                      class="HvVerticalNavigationTreeViewItem-content HvVerticalNavigationTreeViewItem-link er1eqbm0 HvTypography-root HvTypography-body css-1yu9rnx-getStyledComponent-StyledContent e1tnpalo0"
                      href="/?path=/story/structure-vertical-navigation--main"
                      id="hvtreeview2-02-01-02-button"
                      role="button"
                      style="padding-left: 74px;"
                      tabindex="0"
                    >
                      <div
                        class=""
                        data-mui-internal-clone-element="true"
                        style="overflow: hidden; text-overflow: ellipsis; white-space: nowrap;"
                      >
                        HCP Anywhere
                      </div>
                    </a>
                  </li>
                  <li
                    class="HvVerticalNavigationTreeViewItem-node HvVerticalNavigationTreeViewItem-disabled css-1y9s9wc-StyledNode er1eqbm1"
                    id="hvtreeview2-02-01-03"
                  >
                    <div
                      class="HvVerticalNavigationTreeViewItem-content er1eqbm0 HvTypography-root HvTypography-placeholderText css-mbbta7-getStyledComponent-StyledContent e1tnpalo0"
                      id="hvtreeview2-02-01-03-button"
                      role="button"
                      style="padding-left: 74px;"
                      tabindex="0"
                    >
                      <div
                        class=""
                        data-mui-internal-clone-element="true"
                        style="overflow: hidden; text-overflow: ellipsis; white-space: nowrap;"
                      >
                        This Computer
                      </div>
                    </div>
                  </li>
                </ul>
              </li>
            </ul>
          </li>
          <li
            class="HvVerticalNavigationTreeViewItem-node HvVerticalNavigationTreeViewItem-expandable HvVerticalNavigationTreeViewItem-collapsed HvVerticalNavigationTreeViewItem-unselectable css-1y9s9wc-StyledNode er1eqbm1"
            id="hvtreeview2-03"
          >
            <div
              aria-controls="hvtreeview2-03-group"
              aria-expanded="false"
              class="HvVerticalNavigationTreeViewItem-content er1eqbm0 HvTypography-root HvTypography-body css-1yu9rnx-getStyledComponent-StyledContent e1tnpalo0"
              id="hvtreeview2-03-button"
              role="button"
              style="padding-left: 0px;"
              tabindex="0"
            >
              <div
                class="css-f95wjz"
                name="DropDownXS"
              >
                <svg
                  focusable="false"
                  height="12"
                  viewBox="0 0 12 12"
                  width="12"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <path
                    class="color0"
                    d="M5.99994 8.65 1.4 4.05l.7-.7L6 7.25l3.9-3.9.7.7z"
                    fill="var(--uikit-colors-acce1)"
                  />
                </svg>
              </div>
              <div
                class=""
                data-mui-internal-clone-element="true"
                style="overflow: hidden; text-overflow: ellipsis; white-space: nowrap;"
              >
                Administration
              </div>
            </div>
            <ul
              class="HvVerticalNavigationTreeViewItem-group css-zinev9-StyledGroup er1eqbm2"
              id="hvtreeview2-03-group"
            >
              <li
                class="HvVerticalNavigationTreeViewItem-node HvVerticalNavigationTreeViewItem-expandable HvVerticalNavigationTreeViewItem-collapsed HvVerticalNavigationTreeViewItem-unselectable css-1y9s9wc-StyledNode er1eqbm1"
                id="hvtreeview2-03-01"
              >
                <div
                  aria-controls="hvtreeview2-03-01-group"
                  aria-expanded="false"
                  class="HvVerticalNavigationTreeViewItem-content er1eqbm0 HvTypography-root HvTypography-body css-1yu9rnx-getStyledComponent-StyledContent e1tnpalo0"
                  id="hvtreeview2-03-01-button"
                  role="button"
                  style="padding-left: 32px;"
                  tabindex="0"
                >
                  <div
                    class="css-f95wjz"
                    name="DropDownXS"
                  >
                    <svg
                      focusable="false"
                      height="12"
                      viewBox="0 0 12 12"
                      width="12"
                      xmlns="http://www.w3.org/2000/svg"
                    >
                      <path
                        class="color0"
                        d="M5.99994 8.65 1.4 4.05l.7-.7L6 7.25l3.9-3.9.7.7z"
                        fill="var(--uikit-colors-acce1)"
                      />
                    </svg>
                  </div>
                  <div
                    class=""
                    data-mui-internal-clone-element="true"
                    style="overflow: hidden; text-overflow: ellipsis; white-space: nowrap;"
                  >
                    Rest API
                  </div>
                </div>
                <ul
                  class="HvVerticalNavigationTreeViewItem-group css-zinev9-StyledGroup er1eqbm2"
                  id="hvtreeview2-03-01-group"
                >
                  <li
                    class="HvVerticalNavigationTreeViewItem-node HvVerticalNavigationTreeViewItem-selectable HvVerticalNavigationTreeViewItem-unselected css-1y9s9wc-StyledNode er1eqbm1"
                    id="hvtreeview2-03-01-01"
                  >
                    <div
                      class="HvVerticalNavigationTreeViewItem-content er1eqbm0 HvTypography-root HvTypography-body css-1yu9rnx-getStyledComponent-StyledContent e1tnpalo0"
                      id="hvtreeview2-03-01-01-button"
                      role="button"
                      style="padding-left: 74px;"
                      tabindex="0"
                    >
                      <div
                        class=""
                        data-mui-internal-clone-element="true"
                        style="overflow: hidden; text-overflow: ellipsis; white-space: nowrap;"
                      >
                        Log Bundle
                      </div>
                    </div>
                  </li>
                </ul>
              </li>
            </ul>
          </li>
        </ul>
      </nav>
      <div
        class="HvVerticalNavigationActions-root css-btrerb-StyledRoot e11q2fjk0"
      >
        <div
          class="HvVerticalNavigationAction-action e13j4on70 HvTypography-root HvTypography-body css-prxfu3-getStyledComponent-StyledAction e1tnpalo0"
          role="button"
          tabindex="0"
        >
          <div
            class="css-f95wjz"
            name="User"
          >
            <svg
              focusable="false"
              height="16"
              viewBox="0 0 16 16"
              width="16"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                class="color0"
                d="M9.6 8.7A4.49656 4.49656 0 0 0 8.03809 0Q8.019-.00007 8 0a4.481 4.481 0 0 0-4.5 4.46194q-.00009.019 0 .03808A4.563 4.563 0 0 0 6.4 8.7 7.687 7.687 0 0 0 .5 16h1a6.487 6.487 0 0 1 6.474-6.5H8a6.487 6.487 0 0 1 6.5 6.474V16h1a7.493 7.493 0 0 0-5.9-7.3zM4.5 4.5A3.543 3.543 0 0 1 8 1a3.543 3.543 0 0 1 3.5 3.5A3.543 3.543 0 0 1 8 8a3.543 3.543 0 0 1-3.5-3.5z"
                fill="var(--uikit-colors-acce1)"
              />
            </svg>
          </div>
          Profile
        </div>
        <div
          class="HvVerticalNavigationAction-action e13j4on70 HvTypography-root HvTypography-body css-prxfu3-getStyledComponent-StyledAction e1tnpalo0"
          role="button"
          tabindex="0"
        >
          <div
            class="css-f95wjz"
            name="LogOut"
          >
            <svg
              focusable="false"
              height="16"
              viewBox="0 0 16 16"
              width="16"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                class="color0"
                d="M16 16H6v-1h9V1H6V0h10zm-3-7.5v-1H2.7l3-3-.7-.7L.7 8 5 12.2l.7-.7-3-3z"
                fill="var(--uikit-colors-acce1)"
              />
            </svg>
          </div>
          Logout
        </div>
      </div>
    </div>
  </div>
</div>
`;
